src/helpers/StatsClient.ts

```diff
@@ -0,0 +1,176 @@
+import { StatsD, Tags } from 'hot-shots';
+import winston = require('winston');
+
+export interface AddTagsToScopeOptions {
+	scope: string;
+	tags: Tags;
+}
+
+export enum PluginType {
+	ACTIVITY_ANALYZER = 'activity_analyzer',
+	AD_RENDER = 'ad_render',
+	AUDIENCE_FEED = 'audience_feed',
+	BID_OPTIMIZER = 'bid_optimizer',
+	CUSTOM_ACTION = 'custom_action',
+	EMAIL_RENDERER = 'email_renderer',
+	EMAIL_ROUTER = 'email_router',
+	RECOMMENDER = 'recommender',
+}
+
+export enum METRICS_NAME {
+	EXTERNAL_API_CALL_SUCCESS = 'external_api_call_success',
+	EXTERNAL_API_CALL_ERROR = 'external_api_call_error',
+	EXTERNAL_API_CALL_RETRY = 'external_api_call_retry',
+
+	MICS_API_CALL_SUCCESS = 'mics_api_call_success',
+	MICS_API_API_CALL_ERROR = 'mics_api_call_error',
+	MICS_API_API_CALL_RETRY = 'mics_api_call_retry',
+
+	AUDIENCE_FEED_BATCH_UPSERT = 'audience_feed_batch_upsert',
+	AUDIENCE_FEED_BATCH_DELETE = 'audience_feed_batch_delete',
+
+	AD_RENDER_BANNER_GENERATED_DEFAULT = 'ad_render_banner_generated_default',
+	AD_RENDER_BANNER_GENERATED_CUSTOM = 'ad_render_banner_generated_custom',
+
+	ACTIVITY_CREATED = 'activity_created',
+	ACTIVITY_FAILED_TO_CREATE = 'activity_failed_to_create',
+}
+
+export interface InitOptions {
+	/**
+	 * interval to send stats to datadog in ms (default = 10 minutes)
+	 */
+	timerInMs?: number;
+
+	/**
+	 * default plugins tags
+	 */
+	tags: Tags & { pluginType: PluginType; pluginName: string };
+
+	/**
+	 * An optional logger to send Metrics into logs (in debug mode)
+	 */
+	logger?: winston.Logger;
+}
+
+export interface IncrementToStatsOptions {
+	/**
+	 * @example
+	 * ```
+	 * name of your metrics and its value.
+	 * {processed_users: 4}
+	 */
+	metrics: Metrics;
+
+	/**
+	 * the main tag used to filter the metrics in datadog.
+	 * Ex: feedId
+	 */
+	scope?: string;
+}
+
+export interface Metrics {
+	[key: string]: number;
+}
+
+export interface TagsToScope {
+	/**
+	 * Custom tags retrived later on the plugin instantiation.
+	 * Ex: instance context infos.
+	 */
+	[scope: string]: Tags;
+}
+
+interface Stats {
+	[keyOrScope: string]: Metrics | number;
+}
+
+/**
+ * Send stats to datadog
+ */
+export class StatsClient {
+	private static instance: StatsClient;
+	private interval: NodeJS.Timer;
+	private stats: Stats;
+	private client: StatsD;
+	private logger?: winston.Logger;
+	private tagsToScope: TagsToScope;
+
+	private constructor(timerInMs: number, tags: Tags, logger?: winston.Logger) {
+		this.stats = {};
+		this.logger = logger;
+		this.tagsToScope = {};
+		this.client = new StatsD({
+			globalTags: {
+				workerId: process.env.PLUGIN_WORKER_ID || 'DEFAULT_PLUGIN_ID',
```

Comment:
Do we really need the workerId ?

Reply:
I don't now to be honest.

Reply:
We don't need it in my opinion, unless a specific worker fails and we need to know which.

Since everything is dockerised, the environments should be 100% replicable, thus eliminating the need to track with such a fine granularity.